lib/Sema/TypeCheckAttr.cpp

```diff
@@ -3008,8 +3008,29 @@ void AttributeChecker::visitPropertyWrapperAttr(PropertyWrapperAttr *attr) {
 }
 
 void AttributeChecker::visitFunctionBuilderAttr(FunctionBuilderAttr *attr) {
-  // TODO: check that the type at least provides a `sequence` factory?
-  // Any other validation?
+  auto *nominal = dyn_cast<NominalTypeDecl>(D);
+  auto lookupResult = TypeChecker::lookupMember(
+      nominal, nominal->getDeclaredType(),
+      DeclNameRef(D->getASTContext().Id_buildBlock));
+
+  bool hasStaticBuildBlock = false;
+  llvm::SmallVector<ValueDecl *, 4> nonStaticBuildBlocks;
+  for (auto entry : lookupResult.innerResults()) {
+    auto *VD = entry.getValueDecl();
+    if (VD->getInterfaceType()->is<AnyFunctionType>()) {
```

Comment:
I think this should be `isa<FuncDecl>(VD)`. It would be better to invert it and `continue` to reduce nesting.

Reply:
It seems to be valid for `buildBlock` to be a `static var` with function type, which `isa<FuncDecl>(VD)` would disallow, no?

Reply:
Oh right, I forgot that. So I think you need to check for the supported declarations like `VarDecl` and `FuncDecl` ( `ParamDecl` is not necessary as it's a subclass of `VarDecl`).

Reply:
Interesting. The proposal says:

>Calls to function-building methods are type-checked as if a programmer had written `BuilderType.<methodName>(<arguments>)`

and mentions [method signature](https://github.com/apple/swift-evolution/blob/master/proposals/0289-function-builders.md#function-building-methods) of each method and where `buildBlock` takes a single argument with no label. I don't know if zero argument `buildBlock` is valid but Doug can confirm.

Reply:
Here's an even more fun one, that actually makes even the `AnyFunctionType` check erroneous 🙂 

```
struct Callable {
    func callAsFunction(_ exprs: Any...) -> Int { return exprs.count }
}

@_functionBuilder
struct Builder {
    static var buildBlock = Callable()
}

class C {
    @Builder var x: Int {
        1
        2
        3
    }
}
```

Reply:
Latest update removes this check, just verifies that there is *any* static `buildBlock` member. Also added a test for the `Callable` case.